DirectProgramming/DPC++FPGA/ReferenceDesigns/cholesky_inversion/src/cholesky_inversion_demo.cpp

```diff
@@ -0,0 +1,447 @@
+        it Hermitian
+    */
+    for (int mat_idx = 0; mat_idx < kMatricesToInvert; mat_idx++) {
+      int current_matrix = mat_idx * kAMatrixSize;
+
+      // Generate a random matrix R with diagonal elements set to 0
+      // and measure the weights of the off diagonal entries
+      std::vector<T> r, weights, a_matrix_non_hermitian;
+      r.resize(kColumns * kColumns);
+      a_matrix_non_hermitian.resize(kColumns * kColumns);
+      weights.resize(kColumns);
+      for (int row = 0; row < kColumns; row++) {
+        weights[row] = {0};
+        for (int col = 0; col < kColumns; col++) {
+          if (col != row) {
+            int index = (row * kColumns) + col;
+            float random1 = RandomValueInInterval(kRandomMin, kRandomMax);
+            T elem;
+#if COMPLEX == 1
+            float random1I = RandomValueInInterval(kRandomMin, kRandomMax);
+            elem = {random1, random1I};
+            r[index] = elem;
+#else
+            elem = random1;
+            r[index] = elem;
+#endif
+            weights[row] += elem;
+          }
+        }
+
+        // Construct the new diagonal element
+        weights[row] /= kEpsilon;
+        r[(row * kColumns) + row] = weights[row];
+      }
+
+      // Perform the diagonal scaling by solving:
+      // diag(diag(A))*output = A
+      for (int row = 0; row < kColumns; row++) {
+        for (int col = 0; col < kColumns; col++) {
+          int index = row * kColumns + col;
+          a_matrix_non_hermitian[index] = r[index] / r[(row * kColumns) + row];
+        }
+      }
+
+      // Make the matrix Hermitian
+      for (int row = 0; row < kColumns; row++) {
+        for (int col = 0; col < kColumns; col++) {
+          int index = row * kColumns + col;
+
+          a_matrix[current_matrix + index] =
+              (a_matrix_non_hermitian[index] +
+               a_matrix_non_hermitian[(col * kColumns) + row]) /
+              2;
+
+#if COMPLEX == 1
+          if (row > col) {
+            a_matrix[current_matrix + index] =
+                a_matrix[current_matrix + index].conj();
+          }
+#endif
+        }
+      }
+
+#ifdef DEBUG
+      std::cout << "A MATRIX " << mat_idx << std::endl;
+      for (size_t row = 0; row < kRows; row++) {
+        for (size_t col = 0; col < kColumns; col++) {
+          std::cout << a_matrix[current_matrix + (col * kRows) + row] << " ";
+        }  // end of col
+        std::cout << std::endl;
+      }  // end of row
+#endif
+
+    }  // end of mat_idx
+
+    std::cout << "Computing the Cholesky-based inversion of "
+              << kMatricesToInvert << " matri"
+              << (kMatricesToInvert > 1 ? "ces " : "x ") << repetitions
+              << " times" << std::endl;
+
+    CholeskyInversion<T, kComplex>(a_matrix, i_matrix, q, kMatricesToInvert,
+                                   repetitions);
+
+    // For output post-processing (op)
+    T i_matrix_op[kRows][kColumns];
```

Comment:
Can all this verification code go into a subroutine? This `main()` function is getting quite long.